### ISSUE TITLE
This error should not crash the compiler - change return type to Either here

### ISSUE THREAD

**Issue by charlesetc** [carp-lang/Carp#1043]
Hi, thanks for working on carp! I found this bug:

```
$ carp
Welcome to Carp 0.4.2
This is free software with ABSOLUTELY NO WARRANTY.
Evaluate (help) for more information.
> (def lines (String.lines (ref (IO.read-file "/tmp/input.txt"))))
> (def numbers (Array.copy-map &(fn [x] (Maybe.unsafe-from (Int.from-string x))) &lines))
> (Array.copy-map &(fn [x] (Pair.init x ())) &numbers)
carp: I can’t use the type `(Ref Int <StaticLifetime>)` as a member type at no info.

Is it defined and captured in the head of the type definition?. This error should not crash the compiler - change return type to Either here.
CallStack (from HasCallStack):
  error, called at src/Deftype.hs:309:30 in CarpHask-0.4.2.0-K2sdzDEAjn25KRaWCaWwhV:Deftype
```

I'm running ubuntu 20.04. Let me know if there's any more information I can help with!

**Comment by charlesetc** [comment]
Ah maybe I misinterpreted a type error as a compilation error? I'm not sure what "this error should not crash the compiler" means when preceding a stack trace...

**Comment by eriksvedang** [comment]
Sorry about this error, it's more of a note to the compiler writers than to the end user. It hints at an actual compiler error though, you currently can't store refs in structs. But it shouldn't crash the compiler, of course.

**Comment by TimDeve** [comment]
Yes, in that case that's a type error that's a bit to noisy. `copy-map` passes a reference to its function so to store that in a pair you need to do a copy:

```clojure
(Array.copy-map &(fn [x] (Pair.init @x @x )) &numbers)
```

**Comment by eriksvedang** [comment]
Oh wait, I actually misread the error. What @TimDeve said.

**Comment by eriksvedang** [comment]
> Ah maybe I misinterpreted a type error as a compilation error? I'm not sure what "this error should not crash the compiler" means when preceding a stack trace...

It should be fixed. On it.

**Comment by eriksvedang** [comment]
Some investigation shows that we shouldn't have to call `concretizeType` in the deps functions for init (this applies to both sumtypes and structs), they could just return an empty list if there's a problem. 

The misuse should in theory be detected by all the calls to `concretizeType` in Concretize.hs, but for some reason this result is thrown away.

**Comment by eriksvedang** [comment]
More information:

We have a lot of this kind of code in Concretize:

```haskell
        _ <- concretizeTypeOfXObj typeEnv x
        moreCode
```

This works great for getting the *deps* of `x` (which is an effect in a state monad), but the validation is thrown away. To catch that, we need to do something like this:

```haskell
        check <- concretizeTypeOfXObj typeEnv x
        case check of
          Left err -> pure (Left err)
          Right _ -> do
            moreCode
```

When switching to this kind of code, we *do* catch the error in this issue (using refs as members in a struct) but apparently we also have tons of other situations where the current validation raises an error.

For example, when using the `the` annotation `(the (Pair a b) p)` now complains that `a` and `b` are not valid types to instantiate Pair with, and there are tons of similar situations. My current hunch is that the validation should not run when the type being checked is generic – but any feedback on this problem is most welcome!

**Comment by eriksvedang** [comment]
I think one important distinction is whether the check is done on any type in the program, or when defining the type.
